zebrad/src/components/mempool.rs

```diff
@@ -87,48 +104,120 @@ pub struct Mempool {
-            storage: Default::default(),
-            tx_downloads,
+            state: State::Disabled,
             sync_status,
             chain_tip_change,
+            outbound_service,
+            state_service,
+            tx_verifier_service,
+        }
+    }
+
+    /// Update the mempool state (enabled / disabled) depending on how close to
+    /// the tip is the synchronization, including side effects to state changes.
+    fn update_state(&mut self) {
+        // Update enabled / disabled state
+        let is_close_to_tip = self.sync_status.is_close_to_tip();
+        if is_close_to_tip && matches!(self.state, State::Disabled) {
+            let tx_downloads = Box::pin(TxDownloads::new(
+                Timeout::new(self.outbound_service.clone(), TRANSACTION_DOWNLOAD_TIMEOUT),
+                Timeout::new(self.tx_verifier_service.clone(), TRANSACTION_VERIFY_TIMEOUT),
+                self.state_service.clone(),
+            ));
+            self.state = State::Enabled {
+                storage: Default::default(),
+                tx_downloads,
+            };
+        } else if !is_close_to_tip && matches!(self.state, State::Enabled { .. }) {
+            self.state = State::Disabled
+        }
+    }
+
+    /// Return wether the mempool is enabled or not.
+    #[allow(dead_code)]
+    pub fn enabled(&self) -> bool {
+        match self.state {
+            State::Disabled => false,
+            State::Enabled { .. } => true,
         }
     }
 
     /// Get the storage field of the mempool for testing purposes.
     #[cfg(test)]
     pub fn storage(&mut self) -> &mut storage::Storage {
-        &mut self.storage
+        match &mut self.state {
+            State::Disabled => panic!("mempool must be enabled"),
+            State::Enabled { storage, .. } => storage,
+        }
+    }
+
+    /// Get the transaction downloader of the mempool for testing purposes.
+    #[cfg(test)]
+    pub fn tx_downloads(&self) -> &Pin<Box<InboundTxDownloads>> {
+        match &self.state {
+            State::Disabled => panic!("mempool must be enabled"),
+            State::Enabled { tx_downloads, .. } => tx_downloads,
+        }
+    }
+
+    /// Enable the mempool by pretending the synchronization is close to the tip.
+    #[cfg(test)]
+    pub async fn enable(&mut self, recent_syncs: &mut RecentSyncLengths) {
+        use tower::ServiceExt;
+        // Pretend we're close to tip
+        SyncStatus::sync_close_to_tip(recent_syncs);
+        // Wait for the mempool to make it enable itself
+        let _ = self.ready_and().await;
+    }
+
+    /// Disable the mempool by pretending the synchronization is far from the tip.
+    #[cfg(test)]
+    pub async fn disable(&mut self, recent_syncs: &mut RecentSyncLengths) {
+        use tower::ServiceExt;
+        // Pretend we're far from the tip
+        SyncStatus::sync_far_from_tip(recent_syncs);
+        // Wait for the mempool to make it enable itself
+        let _ = self.ready_and().await;
```

Comment:
Possible test bug/unreliability: every `ready_and` should have a `call`.

See my detailed comment on the PR.

```suggestion
        let _ = self.oneshot(Request::TransactionIds).await;
```

Reply:
Done in fe354f51

I'm confused, though. How does this compile? `oneshot` consumes `self`, and here `self` is a `&mut self`

Reply:
So there's a bit of a generics and reference trick here:

`ServiceExt` is implemented for all `Services`:
https://docs.rs/tower/0.4.8/tower/util/trait.ServiceExt.html

`Service` is implemented for all `&mut Service`:
https://docs.rs/tower/0.4.8/tower/trait.Service.html#foreign-impls

So the actual call to `oneshot` desugars to something like:
```rust
ServiceExt::oneshot(self: &mut Mempool, req: mempool::Request)
```

The method call moves the `&mut Mempool` reference into `oneshot`, making it unusable in the rest of the method. (But there are no further uses of `self` in the method, so that's ok.)

Instead, the `&mut`'s exclusive ownership reverts to the caller once the awaited future returns.

(If it was a shared reference, it would be copied, not moved. If it was owned, it would be moved, and the tail function would `Drop` it before returning.)

Reply:
Cool, thanks for the great explanation :grin: